pkg/controllers/user/multiclusterapp/controller.go

```diff
@@ -0,0 +1,267 @@
+	}
+	appName := temp[0]
+
+	// for all targets, create the App{} instance, so that helm controller App lifecycle can pick it up
+	// only one app per project named mcapp-{{mcapp.Name}}
+	appNameInProject := multiClusterAppPrefix + appName
+	targetNamespace := appNameInProject
+	for _, t := range mcapp.Spec.Targets {
+		split := strings.SplitN(t.ProjectName, ":", 2)
+		if len(split) != 2 {
+			return nil, fmt.Errorf("error in splitting project ID")
+		}
+		projectNS := split[1]
+		// check if the target project in this iteration is same as the cluster in current context, if not then don't create namespace and app else it
+		// will be in the wrong cluster
+		if split[0] != m.ClusterName {
+			logrus.Info("Not for the current cluster")
+			continue
+		}
+
+		// check if this app already exists
+		a, err := m.AppClient.GetNamespaced(projectNS, appNameInProject, metav1.GetOptions{})
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				return nil, fmt.Errorf("error checking if app already exists in project")
+			}
+			// err is IsNotFound
+		} else if a != nil && a.Name == appNameInProject {
+			logrus.Infof("App for multiclusterapp %v already exists in project %v", mcapp.Name, t.ProjectName)
+			continue
+		}
+
+		app := pv3.App{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        appNameInProject,
+				Namespace:   projectNS,
+				Annotations: ann,
+				Labels:      set,
+			},
+			Spec: pv3.AppSpec{
+				ProjectName:         t.ProjectName,
+				TargetNamespace:     targetNamespace,
+				ExternalID:          externalID,
+				MultiClusterAppName: mcapp.Name,
+			},
+		}
+
+		if ans, ok := answerMap[t.ProjectName]; ok {
+			app.Spec.Answers = ans
+			answerFound = true
+		}
+
+		if !answerFound {
+			// find the answers for the cluster of this project
+			split := strings.SplitN(t.ProjectName, ":", 2)
+			if len(split) != 2 {
+				return mcapp, fmt.Errorf("error in splitting project name")
+			}
+			clusterName := split[0]
+			if ans, ok := answerMap[clusterName]; ok {
+				app.Spec.Answers = ans
+				answerFound = true
+			}
+
+			if !answerFound {
+				if ans, ok := answerMap[globalScopeAnswersKey]; ok {
+					app.Spec.Answers = ans
+				}
+			}
+		}
+
+		// Create the target namespace first
+		// Check if the target namespace already exists, if not then continue, if it does then add a random 5-character string at the end to make it unique
+		// Adding the projectId as an annotation is necessary, else the API/UI and UI won't list any of the resources from this namespace
+		nsLabel := make(map[string]string)
+		nsLabel[projectIDFieldLabel] = projectNS
+		nsAnnotation := make(map[string]string)
+		nsAnnotation[projectIDFieldLabel] = t.ProjectName
+		nsAnnotation[creatorIDAnn] = creatorID
+		n := v1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        targetNamespace,
+				Labels:      nsLabel,
+				Annotations: nsAnnotation,
+			},
+		}
+		_, err = m.NsClient.Get(targetNamespace, metav1.GetOptions{})
```

Comment:
I feel like we want to handle the namespace creation before generating the app

Reply:
So we are creating namespace before calling Apps.Create

Reply:
do you think we can extract app creation based on target, to its own method? instead of keeping a big code block in if/else method